File: java/src/main/java/com/ellen/java/jmvp/base/BaseMvpActivity.java
package com.ellen.java.jmvp.base;

import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import javax.inject.Inject;

public abstract class BaseMvpActivity<P extends BasePresenter> extends AppCompatActivity {

    @Inject
    protected P mPresenter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        initMvp();

        //调用MVP的生命周期函数attach
        mPresenter.attachView();
        mPresenter.mView.attachPresenter(mPresenter);
        mPresenter.mModel.attachPresenter(mPresenter);

        mPresenter.attachLifeCycle(getLifecycle());
    }

    protected abstract void initMvp();
}

File: app/src/main/java/com/ellen/androidarchitecture/kmvp/base/BaseMvpActivity.kt
package com.ellen.androidarchitecture.kmvp.base

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import javax.inject.Inject

abstract class BaseMvpActivity<P : BasePresenter<*,*>> : AppCompatActivity(),BaseView {
    @Inject
    protected lateinit var mPresenter:P

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initMVP()

        //调用MVP的绑定attach周期函数
        mPresenter.attachedByView()
        mPresenter.mView.attachedByPresenter()
        mPresenter.mModel.attachedByPresenter()

        //让Presenter感知生命周期，以防止内存泄漏
        mPresenter.attachLifecycle(lifecycle)
    }

    abstract fun initMVP();
}